util/trace/src/collections/hash_map.rs

```diff
@@ -0,0 +1,159 @@
+use ckb_logger::trace;
+use std::{
+    borrow::Borrow,
+    collections::{
+        self as base,
+        hash_map::{IntoIter, Iter, IterMut, RandomState},
+    },
+    fmt,
+    hash::{BuildHasher, Hash},
+    ops,
+    time::Instant,
+};
+
+#[derive(Clone)]
+pub struct HashMap<K, V, S = RandomState> {
+    tag: String,
+    base: base::HashMap<K, V, S>,
+    next_check: Instant,
+}
+
+impl<K, V, S> From<HashMap<K, V, S>> for base::HashMap<K, V, S> {
+    fn from(map: HashMap<K, V, S>) -> Self {
+        map.base
+    }
+}
+
+impl<K, V, S> ops::Deref for HashMap<K, V, S> {
+    type Target = base::HashMap<K, V, S>;
+    fn deref(&self) -> &Self::Target {
+        &self.base
+    }
+}
+
+impl<K, V, S> ops::DerefMut for HashMap<K, V, S> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.base
+    }
+}
+
+impl<K, V, S> HashMap<K, V, S> {
+    pub(crate) fn trace_size(&mut self) {
+        let now = Instant::now();
+        if now >= self.next_check {
+            self.next_check = now + crate::trace_interval();
+            trace!(
+                "HashMap {{ tag: {}, len: {}, capacity: {} }}",
+                &self.tag,
+                self.len(),
+                self.capacity(),
+            );
+        }
+    }
+
+    pub fn tag(&self) -> &str {
+        &self.tag
+    }
+}
+
+impl<K: Hash + Eq, V> HashMap<K, V, RandomState> {
+    pub fn new(tag: String) -> Self {
+        let now = Instant::now();
+        Self {
+            tag,
+            base: base::HashMap::new(),
+            next_check: now + crate::trace_interval(),
+        }
+    }
+}
+
+impl<K, V, S> HashMap<K, V, S>
+where
+    K: Eq + Hash,
+    S: BuildHasher,
+{
+    pub fn with_capacity_and_hasher(
+        tag: String,
```

Comment:
Two solutions come to my mind to avoid passing tag via the constructor.

1. Generate the tag from [backtrace](https://github.com/rust-lang/backtrace-rs)
2. Use a global variable to set the tag in a thread local variable. E.g.

```
thead_local trace_labels_stack = [];

def foo():
  trace_labels_stack.push("bar1")
  bar()
  trace_labels_stack.pop()

  trace_labels_stack.push("bar2")
  bar()
  trace_labels_stack.pop()

def bar()
  trace_labels_stack.push("inner")
  inner = HashMap.new()
  trace_labels_stack.pop()
```

Reply:
> 1. Generate the tag from backtrace

I tried. It's difficult to generate a brief and explicit tag from backtrace.
- If keep the full backtrace, the outputs will be too much.
- If keep only one frame, it's difficult to select which frame is enough to point out the key location. 

> 2. Use a global variable to set the tag in a thread local variable. 

I will try this.